FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+) -> Result<[u8; 32]>;
+```
+
+#### Namespace: `send`
+
+```rust
+/// Sends a message to another actor, and returns the exit code and block ID of the return
+/// result.
+pub fn send(
+    recipient_off: *const u8,
+    recipient_len: u32,
+    method: u64,
+    params: u32,
+    value_hi: u64,
+    value_lo: u64,
+) -> Result<Send>;
+
+#[repr(C)]
+pub struct Send {
+    pub exit_code: u32,
+    pub return_id: BlockId,
+}
+```
+
+#### Namespace: `self`
+
+```rust
+/// Gets the current root for the calling actor.
+///
+/// If the CID doesn't fit in the specified maximum length (and/or the length is 0), this
+/// function returns the required size and does not update the cid buffer.
+pub fn root(cid: *mut u8, cid_max_len: u32) -> Result<u32>;
+
+/// Sets the root CID for the calling actor. The new root must be in the reachable set.
+pub fn set_root(cid: *const u8) -> Result<()>;
+
+/// Gets the current balance for the calling actor.
+pub fn current_balance() -> Result<TokenAmount>;
+
+/// Destroys the calling actor, sending its current balance
+/// to the supplied address, which cannot be itself.
+pub fn self_destruct(addr_off: *const u8, addr_len: u32) -> Result<()>;
+```
+
+#### Namespace: `vm`
+
+```rust
+/// Abort execution with the given code and message. The code is recorded in the receipt, the
+/// message is for debugging only.
+pub fn abort(code: u32, message: *const u8, message_len: u32) -> !;
+```
+
+### Externs
+
+```rust
+/// Consensus related methods.
+pub trait Consensus {
+    /// Verify a consensus fault.
+    fn verify_consensus_fault(
+        &self,
+        h1: &[u8],
+        h2: &[u8],
+        extra: &[u8],
+    ) -> Result<Option<ConsensusFault>>;
+}
+
+/// Randomness provider trait
+pub trait Rand {
+    /// Gets 32 bytes of randomness for ChainRand paramaterized by the DomainSeparationTag,
+    /// ChainEpoch, Entropy from the ticket chain.
+    fn get_chain_randomness(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    fn get_chain_randomness_looking_forward(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    /// Gets 32 bytes of randomness for ChainRand paramaterized by the DomainSeparationTag,
+    /// ChainEpoch, Entropy from the latest beacon entry.
+    fn get_beacon_randomness(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+
+    fn get_beacon_randomness_looking_forward(
+        &self,
+        pers: DomainSeparationTag,
+        round: ChainEpoch,
+        entropy: &[u8],
+    ) -> Result<[u8; 32]>;
+}
+
+```
+
+### IPLD memory model
+
+IPLD is the data model of Filecoin. Actors can be construed as logic that
+receives an input IPLD graph, performs computation, and (optionally) mutates the
+IPLD graph by saving an alternative root in the actor's state.
```

Comment:
nit: "and returns a new IPLD graph".

I'd avoid the word "mutate" as IPLD is copy-on-write.

Reply:
Works for me, I had worded it that way before, but it's not technically accurate today. However, it is conceptually correct.